Upgrade all the way to Rails 6.0.3.4

t27duck opened the pull request:
Closes #80 

## The Problem

The app is on Rails 5.1.x which is no longer supported. It was asked that the app be up to date with Rails 6.0.x.

## The Solution

Over multiple commits, the app has been stepped through "upgrading" from the Rails 5.1.x it was on, to the last 5.1.x, to the last 5.2.x, and then to the latest 6.0.x.

A lot of gems had to be updated to accommodate the update.

At least on my system, specs pass, cucumber tests "function" (see my PR comment in the Gemfile), and the local dev server renders the app. Was able to log in as the local admin and "approve" job postings.

The app itself could probably use a lot more love (like, remove coffescript gem, maybe go to webpacker, stop with cucumber, revisit and update the config files, etc etc), but this at least gets it to a current Rails?

t27duck commented on Gemfile:
```diff
@@ -54,7 +54,7 @@ group :test do
+  gem 'factory_bot_rails'
   gem 'faker'
   gem 'fakeweb'
   gem 'capybara'
```
A lot of muddy work regarding cucumber, capybara, and related gems here... there's actually a downgrade to one of them in this change set. Due to major gherkin changes along with capybara matchers and other "stuff", someone more knowledgeable than I would need to more or less audit and/or rewrite the feature tests. For now, this works, minus what SHOULD be a safe "error" where minitest is invoked accidentally and complains about an unknown `--format` flag.

t27duck commented on Gemfile:
```diff
@@ -1,19 +1,19 @@
-gem 'haml', '5.0.4'
-gem 'sass'
+gem 'haml', '~> 5.1.2'
+gem 'sassc'
```
The `sass` gem is dead. `sassc` is the last hold out and needed for what's left of asset pipeline.

t27duck commented on app/assets/javascripts/application.js:
```diff
@@ -3,7 +3,7 @@
 //= require jquery.easing
 //= require jquery.blockUI
-//= require jquery.elastic.source
+//= require jquery.elastic
```
Sometime asset pipeline apparently was updated so it doesn't like files that end in "source"? Changing this filename made it suddenly find it again in current versions of the gem.

t27duck commented on app/controllers/admin/jobs_controller.rb:
```diff
@@ -28,6 +28,6 @@ def resource
 
   def job_params
-    params[:job].permit(:title, :description, :published_at)
+    params.require(:job).permit(:title, :description, :company, :published_at)
```
So `company` was never here. Meaning old Rails was either letting it in or `company` was never being set for the records?

wuz replied:
Gonna guess the former since we use the company name in the job titles.

t27duck commented on app/models/job.rb:
```diff
@@ -36,7 +36,27 @@ def name
+    will_save_change_to_title? || will_save_change_to_company?
+  end
+
+  # Compatibility method for slugged gem due to current Rails changes to the "ActiveModel::Dirty" methods
```
I don't know what the "slugged" gem is or why a full on gem was needed for pretty much a `to_param` replacement. All I know is, I hate it.

Anyway, these dynamic methods are no longer in Rails and have replacements. But, the slugged gem hasn't been touched in six years so... yeah.

rreinhardt9 replied:
Oh nice!!! Last time I looked into this the "slugged" gem is where I got stuck. like you said, it hadn't been updated in years. I think it's being used to create the slugs in the links for jobs. Love seeing how you got passed that part!

wuz replied:
Been looking at swapping the slugged gem out for something like [`friendly_id`](https://github.com/norman/friendly_id) here. Might be something to add as a new issue :D

t27duck commented on config/application.rb:
```diff
@@ -27,6 +27,6 @@ class Application < Rails::Application
     ]
     config.assets.debug = true
-    config.cache_store = :dalli_store
+    config.cache_store = :mem_cache_store
```
This is the same compatibility-wise. Rails more or less just absorbed the adapter.

t27duck commented on config/environments/production.rb:
```diff
@@ -63,7 +63,7 @@
   # Enable locale fallbacks for I18n (makes lookups for any locale fall back to
   # the I18n.default_locale when a translation can not be found)
-  config.i18n.fallbacks = true
+  config.i18n.fallbacks = [I18n.default_locale]
```
Current day `I18n` installs warn you that this needs to change or things "may" break depending on how much this app relies on localizations.

t27duck commented on config/initializers/assets.rb:
```diff
@@ -1 +1 @@
-Rails.application.config.assets.precompile += %w( showdown.js )
+Rails.application.config.assets.precompile += %w( showdown.js application.css application.js indyhackers-logo.png )
```
All I know is, current asset pipeline wouldn't work without this.

t27duck commented on config/initializers/secret_token.rb:
```diff
@@ -1,5 +1,5 @@
 if Rails.env.production?
-  IndyhackersRails::Application.config.secret_token = ENV['OMG_SECRET_TOKEN']
+  IndyhackersRails::Application.config.secret_key_base = ENV['OMG_SECRET_TOKEN']
```
This may or may not work in current Rails? Usually it looks for a `RAILS_SECRET_KEY_BASE` or some environment variable named similarly. I **think** this will still work.

t27duck commented on config/initializers/secret_token.rb:
```diff
@@ -1,5 +1,5 @@
+  IndyhackersRails::Application.config.secret_key_base = ENV['OMG_SECRET_TOKEN']
 else
-  IndyhackersRails::Application.config.secret_token = '4ef7f3c920e2aade871b21082bb5ad5fadf5e35bd9b0997f05c2705d2891f658c3f88565ca6d0232ea224c68a5a3d50c3d73089deed21b9eee49f06058fd6b71'
+  IndyhackersRails::Application.config.secret_key_base = '4ef7f3c920e2aade871b21082bb5ad5fadf5e35bd9b0997f05c2705d2891f658c3f88565ca6d0232ea224c68a5a3d50c3d73089deed21b9eee49f06058fd6b71'
```
Technically not needed anymore. Rails 6 will set the secret key for you in dev and test mode. But, I suppose out of scope for this PR.

t27duck commented on config/newrelic.yml:
```diff
@@ -1,4 +1,4 @@
- production:
+production:
```
Uh, yeah. Did this ever work?

t27duck commented on spec/factories.rb:
```diff
@@ -1,10 +1,10 @@
   factory :job do
-    title Faker::Lorem.words.join(' ')
-    description Faker::Lorem.paragraphs.join
+    title { Faker::Lorem.words.join(' ') }
```
Factories require the attributes be in callable blocks from now on.

t27duck commented:
Updated to the most recent Rails that was released today. Should be good.

t27duck commented:
Meant to do this sooner... per the discussion in the slack, I'll ping @armilam as a backup eventual looker-overer if someone else can't get to it.

wuz commented on app/models/redirect.rb:
```diff
@@ -12,6 +12,6 @@ def url_is_valid
 
   def visit!
-    update_attributes!(last_visited_at: Time.now, :visits => visits + 1)
+    update!(last_visited_at: Time.now, :visits => visits + 1)
```
I'm not positive what we do with "visits", but I'm wondering if it might be beneficial to move this over to a real analytics platform? Maybe something self-hosted and not GA? @mileszs and @armilam what do you think?

armilam replied:
I have no idea what visits is here. I'm not sure what we want to do with visit data, but if there is a purpose, then I agree a real platform might be in order if we can get it free or super cheap.

wuz commented on features/step_definitions/admin_steps.rb:
```diff
@@ -1,9 +1,9 @@
 Given /^I am signed up as "([^\"]*)"$/ do |email|
-  FactoryGirl.create(:admin, :email => email, :password => "rockstar")
+  FactoryBot.create(:admin, :email => email, :password => "rockstar")
```
Oh thank god we are using FactoryBot now lol

wuz reviewed (approved):
This looks pretty great to me, but I'll let @armilam weigh in before we merge the change! Thanks so much for doing this!

armilam commented:
After scanning the code the changes seem fine. I trust you and @wuz's review. I'm just going to run this locally and do some smoke testing real quick.

armilam commented:
http://localhost:3000/about

![Screen Shot 2020-10-08 at 2 33 52 PM](https://user-images.githubusercontent.com/430986/95499461-4c3eb880-0973-11eb-8dac-688f3b491ccd.png)

t27duck commented:
Can you guess what page I didn't click on? 🙃 

I'll see what I can do.

armilam commented:
Thanks! Besides that one thing, it appears everything works. Thank you for getting us up to date!

t27duck commented:
@armilam The about us page should be "fixed." By "fixed" I mean, "The board member images were slip between `public/images` and `app/assets/images` so I took the easy way out and moved them all out of asset pipeline to `public/images/board` (for organizational purposes), adjusted the `image_tag` and it all worked locally for me."

I did an `ag` on the codebase to make sure those images weren't referenced anywhere else. Looks like they weren't.

armilam commented:
It looks like everything is working. @t27duck thank you for doing this - I really appreciate it!

Now to test in production. :yolo:
